### ISSUE TITLE
Questions about TensorRT and Pytorch inference speed

### ISSUE THREAD

**Issue by ponytaill** [ultralytics/ultralytics#16920]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi, I followed the [tutorials](https://docs.ultralytics.com/integrations/tensorrt/?h=tensorrt) to export yolo11s to tensorrt and validated, but got unusual results that tensorrt was slower than pytorch. 
Here is my code and results.(By the way, I use KITTI dataset to train yolo11s, so the val data is from KITTI)

from ultralytics import YOLO
import torch
model = YOLO("best.pt")
torch.cuda.reset_max_memory_allocated()
metrics = model.val(data="/root/autodl-tmp/datasets/kitti.yaml")
if torch.cuda.is_available():
    print(f"Peak GPU Memory Allocated: {torch.cuda.max_memory_allocated() / 1024 ** 2:.2f} MB")
    print(f"Peak GPU Memory Reserved: {torch.cuda.max_memory_reserved() / 1024 ** 2:.2f} MB")
model.export(format="engine", half=True) 

I got:
![图片](https://github.com/user-attachments/assets/5c9f324b-867f-46db-a493-e732e9d13418)

tensorrt_model = YOLO("best.engine")
torch.cuda.reset_max_memory_allocated()
metrics = tensorrt_model.val(data="/root/autodl-tmp/datasets/kitti.yaml")

if torch.cuda.is_available():
    print(f"Peak GPU Memory Allocated: {torch.cuda.max_memory_allocated() / 1024 ** 2:.2f} MB")
    print(f"Peak GPU Memory Reserved: {torch.cuda.max_memory_reserved() / 1024 ** 2:.2f} MB")

 I got:
![图片](https://github.com/user-attachments/assets/7d24d518-399f-40d7-a616-c1c2cedff71e)

As you can see, the pytorch speed is **0.1ms preprocess, 0.7ms inference, 0.0ms loss, 0.7ms postprocess** per image, but tensorrt speed is **0.5ms preprocess, 1.0ms inference, 0.0ms loss, 1.3ms postprocess** per image .

I have seen some similar issues, and the suggested solution is to add half=true. However I have met the condition, the speed is still slow, why? I'd appreciate it if anyone could help me.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ponytaill, thank you for bringing this to our attention 🚀!

We recommend reviewing the [Docs](https://docs.ultralytics.com) where many common questions, including performance comparisons, might already be addressed. For your specific setup with TensorRT and PyTorch, ensuring you have followed the correct export steps is crucial.

For 🚀 Bug Reports, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) if not already included. This helps us better understand and debug the issue.

Since you're encountering performance discrepancies, it may be helpful also to check if your environment meets all the specified requirements, and try running the latest version of the library:

```bash
pip install -U ultralytics
```

If your question pertains to custom setups or config tweaks ❓, sharing more details such as dataset examples and training logs will be beneficial. Please confirm you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join our vibrant Ultralytics community for more interactive discussions. We have real-time chat on [Discord](https://ultralytics.com/discord) 🎧, long-form discussions on [Discourse](https://community.ultralytics.com), and a community Subreddit for sharing insights and solutions.

This is an automated response. An Ultralytics engineer will review your issue and provide further assistance soon. 😊

**Comment by Y-T-G** [comment]
1. `model.val()` runs at batch 16 by default. So your Pytorch model is using batching, but your TensorRT model isn't. Export your TensorRT model with `batch=16` and run `model.val(..., batch=16)`

But batched inference time is not the same as single image inference time.

2. PyTorch model doesn't use a fixed image size while the TensorRT model uses a fixed image size.

Here's, the PyTorch inference.

```python
In [11]: model = YOLO("yolov8n.pt")

In [12]: results = model([ASSETS / "bus.jpg"])

0: 640x480 4 persons, 1 bus, 1 stop sign, 33.3ms
Speed: 2.9ms preprocess, 33.3ms inference, 1.6ms postprocess per image at shape (1, 3, 640, 480)
```

It resized to 480x640.

Here's, the TensorRT inference:

```python
In [15]: results = model([ASSETS / "bus.jpg"])

0: 640x640 4 persons, 1 bus, 3.2ms
Speed: 2.0ms preprocess, 3.2ms inference, 1.5ms postprocess per image at shape (1, 3, 640, 640)
```

It resized to 640x640. The PyTorch model would be faster because it didn't use the full image size.


So TensorRT is faster. But your benchmarking is not correct.

**Comment by ponytaill** [comment]
Hi, @Y-T-G. Thanks for your reply! I try to export my TensorRT model with batch=16 and run model.val(..., batch=16), however I meet a problem:
[TRT] [E] IExecutionContext::setInputShape: Error Code 3: API Usage Error (Parameter check failed, condition: engineDims.d[i] == dims.d[i]. Static dimension mismatch while setting input shape for images. Set dimensions are [13,3,640,640]. Expected dimensions are [16,3,640,640].)
It seems that the input of TensorRT model has being fixed and error occurs when the data size is not evenly divided by a batch of 16 sizes. I didn't find a good solution on the Internet. Do you have any suggestions?

**Comment by Y-T-G** [comment]
It's probably during the last iteration. You can try deleting 13 val images from val folder.

**Comment by ponytaill** [comment]
Thanks! In response to your second point above, if I want the pytorch model to have the same input size as the tensorrt model, do I need to change the source code? Is there an easy way to do this? I considered changing all the image sizes to (640,640), but that didn't seem right.

**Comment by Y-T-G** [comment]
You will have to modify the source code and set this to False.

https://github.com/ultralytics/ultralytics/blob/15e6133534d82ec6845308f1e258d77c1b9bab36/ultralytics/data/augment.py#L1527

**Comment by ponytaill** [comment]
Weird. After I modify the code, the results don't seem to have changed. The first one below is the result of the Pytorch model, and the second one is TensorRT, where you can see that the size of the individual images is different.
![1](https://github.com/user-attachments/assets/9e14dee3-a839-4dae-8666-ce5b66e55ee0)
![2](https://github.com/user-attachments/assets/23d35beb-7bc5-4f9a-997c-6cdba7d0ee6d)

**Comment by Y-T-G** [comment]
How did you modify it? In the installed package file?

**Comment by ponytaill** [comment]
> How did you modify it? In the installed package file?

In`/root/miniconda3/envs/yolo/lib/python3.10/site-packages/ultralytics/data`

**Comment by Y-T-G** [comment]
What size does it output when you run this?

```python
from ultralytics import YOLO, ASSETS

model = YOLO("yolo11n.pt")
results = model(ASSETS / "bus.jpg") 
```

**Comment by Y-T-G** [comment]
You will have to pass `rect=False` to `model.val()`

**Comment by glenn-jocher** [comment]
You're welcome! If you have any more questions, feel free to ask.

**Comment by Petros626** [comment]
@ponytaill which evaluation metrics you're using for KITTI?

**Comment by ponytaill** [comment]
> @ponytaill which evaluation metrics you're using for KITTI?

Sorry for the late reply. I use the default mAP.

**Comment by Petros626** [comment]
> > @ponytaill which evaluation metrics you're using for KITTI?
> 
> Sorry for the late reply. I use the default mAP.

But normally Ap50 and Ap70 is the standard related to scientific papers. Did you evaluated with ultralytics or another kitti eval tool using a tensorrt model?

**Comment by ponytaill** [comment]
> > > @ponytaill which evaluation metrics you're using for KITTI?
> > 
> > 
> > Sorry for the late reply. I use the default mAP.
> 
> But normally Ap50 and Ap70 is the standard related to scientific papers. Did you evaluated with ultralytics or another kitti eval tool using a tensorrt model?

Sorry, I did not.

**Comment by glenn-jocher** [comment]
Thank you for clarifying. If you'd like to evaluate AP50 or AP70 directly, you can modify the YOLO evaluation configurations or use a dedicated KITTI evaluation tool. Let us know if you need guidance on setting this up!
